File: Chains/Waveshine.cpp
#include <cmath>
#include <cstdlib>

#include "../Util/Constants.h"
#include "Waveshine.h"

void Waveshine::PressButtons()
{
    //If we're falling off the stage, then we must have slid off. Jump back on.
    //(Also, make sure that we're not trying to jump two frames in a row. If so, it would be interpreted as one jump)
    if(m_state->m_memory->player_two_action == FALLING &&
        std::abs(m_state->m_memory->player_two_x) > m_state->getStageEdgeGroundPosition() &&
        m_state->m_memory->frame != m_frameJumped+1)
    {
        bool onRight = m_state->m_memory->player_two_x > 0;
        m_frameJumped = m_state->m_memory->frame;
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, onRight ? 0 : 1, .5);
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    //If they're hurt and we're chasing them. Then don't try to pivot. Go into a running state then shine
    if(m_frameShined == 0 &&
        m_state->m_memory->player_two_action == DASHING &&
        m_state->isDamageState((ACTION)m_state->m_memory->player_one_action) &&
        m_state->m_memory->player_one_hitstun_frames_left > (FOX_DASH_FRAMES - m_state->m_memory->player_two_action_frame))
    {
        bool onRight = m_state->m_memory->player_one_x < m_state->m_memory->player_two_x;
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, onRight ? 0 : 1, .5);
        return;
    }

    //Avoid moonwalking
    if(m_state->m_moonwalkRisk)
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, .5);
        m_moonwalkPrevent = true;
        return;
    }
    if(m_moonwalkPrevent)
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, .5);
        m_moonwalkPrevent = false;
        return;
    }

    //Pivot shine! You can't shine from a dash animation. So make it a pivot
    if(m_frameShined == 0 &&
        m_state->m_memory->player_two_action == DASHING)
    {
        bool facingRight = m_state->m_memory->player_two_facing;
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, facingRight ? 0 : 1, .5);
        return;
    }

    //Do nothing if we're in hitlag
    if(m_state->m_memory->player_two_hitlag_frames_left > 0)
    {
        if(m_hitlagFrames == 0)
        {
            m_hitlagFrames = m_state->m_memory->player_two_hitlag_frames_left;
        }
        m_controller->emptyInput();
        return;
    }

    if(m_frameShined == 0)
    {
        //Shine
        m_isBusy = true;
        m_frameShined = m_state->m_memory->frame;
        m_controller->pressButton(Controller::BUTTON_B);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
        return;
    }

    if(m_state->m_memory->frame == m_frameShined+m_hitlagFrames+1)
    {
        //Let go of down b
        m_controller->releaseButton(Controller::BUTTON_B);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, .5);
        return;
    }

    if(m_state->m_memory->frame == m_frameShined+m_hitlagFrames+4)
    {
        //Jump out of our shine
        m_frameJumped = m_state->m_memory->frame;
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    if(m_state->m_memory->frame == m_frameJumped+1)
    {
        m_controller->releaseButton(Controller::BUTTON_Y);
        return;
    }

    if(m_state->m_memory->player_two_action == KNEE_BEND &&
        m_frameKneeBend == 0)
    {
        m_frameKneeBend = m_state->m_memory->frame;
        return;
    }

    //If we're in the air, there won't be any knee bending frames
    if(m_state->m_memory->frame == m_frameKneeBend+1 ||
        m_state->m_memory->player_two_action == JUMPING_ARIAL_FORWARD ||
        m_state->m_memory->player_two_action == JUMPING_ARIAL_BACKWARD)
    {
        m_airdodgeFrame = m_state->m_memory->frame;
        m_controller->pressButton(Controller::BUTTON_L);
        //TODO: still assumes we're facing the opponent
        //If we're close to the edge, don't wavedash off
        if(std::abs(m_state->m_memory->player_two_x) + 2 > m_state->getStageEdgeGroundPosition())
        {
            //If we're super duper close the the edge, we HAVE to wavedash back, or die
            if(m_state->m_memory->player_two_x > 0)
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .2, .2);
            }
            else
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .8, .2);
            }
        }
        //If we're only kinda close to the edge, just wavedash downward
        else if(std::abs(m_state->m_memory->player_two_x) + 18 > m_state->getStageEdgeGroundPosition())
        {
            m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
        }
        else
        {
            if(m_state->m_memory->player_one_x > m_state->m_memory->player_two_x)
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .8, .2);
            }
            else
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .2, .2);
            }
        }
        return;
    }

    if(m_state->m_memory->frame == m_airdodgeFrame+1)
    {
        m_isBusy = false;
        m_controller->releaseButton(Controller::BUTTON_L);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, .5);
        return;
    }
}

bool Waveshine::IsInterruptible()
{
    //If the enemy is off the stage, let's quit early
    if(std::abs(m_state->m_memory->player_one_x) > m_state->getStageEdgeGroundPosition() + .001)
    {
        return true;
    }

    if(m_state->m_memory->frame - m_startingFrame > 20)
    {
        return true;
    }
    return !m_isBusy;
}

Waveshine::Waveshine()
{
    m_startingFrame = m_state->m_memory->frame;
    m_frameJumped = 0;
    m_frameShined = 0;
    m_hitlagFrames = 0;
    m_frameKneeBend = 0;
    m_airdodgeFrame = 0;
    m_isBusy = false;
    m_moonwalkPrevent = false;
}

Waveshine::~Waveshine()
{
}

File: Chains/waveshine.py
import melee
from melee.enums import Action, Button
from Chains.chain import Chain

# Shine, then wavedash
class Waveshine(Chain):
    # Distance argument is a multiplier to how far we'll wavedash
    # 0 is straight in place
    # 1 is max distance
    def __init__(self, distance=1):
        self.hasshined = False
        self.distance = distance

    def step(self, gamestate, smashbot_state, opponent_state):
        controller = self.controller

        shineablestates = [Action.TURNING, Action.STANDING, Action.WALK_SLOW, Action.WALK_MIDDLE, \
            Action.WALK_FAST, Action.EDGE_TEETERING_START, Action.EDGE_TEETERING, Action.CROUCHING, \
            Action.RUNNING, Action.RUN_BRAKE, Action.CROUCH_START, Action.CROUCH_END] # Added last 3 because Smashbot would do nothing if Chains.Waveshine was called during them

        jcshine = (smashbot_state.action == Action.KNEE_BEND) and (smashbot_state.action_frame == 3)
        lastdashframe = (smashbot_state.action == Action.DASHING) and (smashbot_state.action_frame == 12)

        # If somehow we are off stage, give up immediately
        if smashbot_state.off_stage:
            self.interruptible = True
            controller.empty_input()
            return

        # Shine clank! We should shine again if we're in range
        if not opponent_state.hitlag and opponent_state.hitstun_frames_left == 0 and \
                smashbot_state.action == Action.SWORD_DANCE_2_MID_AIR and \
                gamestate.distance < 11.8:
            self.hasshined = False

        # Do the shine if we can
        if not self.hasshined and ((smashbot_state.action in shineablestates) or lastdashframe or jcshine):
            self.interruptible = False
            controller.press_button(Button.BUTTON_B)
            controller.tilt_analog(Button.BUTTON_MAIN, .5, 0)
            self.hasshined = True
            return

        # If we're in the early knee bend frames, just hang on and wait
        if (smashbot_state.action == Action.KNEE_BEND) and (smashbot_state.action_frame < 3):
            controller.empty_input()
            return

        # Pivot. You can't shine from a dash animation. So make it a pivot
        if smashbot_state.action == Action.DASHING:
            # Turn around
            self.interruptible = True
            controller.release_button(Button.BUTTON_B)
            controller.tilt_analog(Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
            #controller.press_button(Button.BUTTON_Y) #attempt JC shine instead of pivot shine
            return

        # In the off-chance waveshine.py gets called during GRAB_WAIT, down-throw
        if smashbot_state.action == Action.GRAB_WAIT:
            controller.tilt_analog(Button.BUTTON_MAIN, .5, 0)
            if self.controller.prev.main_stick[1] == 0:
                controller.empty_input()
            return

        isInShineStart = smashbot_state.action in [Action.DOWN_B_GROUND_START, Action.DOWN_B_GROUND]
        needsJC = smashbot_state.action in [Action.SHIELD_RELEASE, Action.SHIELD, Action.TURNING_RUN] #Added TURNING_RUN in case waveshine gets called during that animation

        # Jump out of shield, turning run, or tilt turn
        if needsJC or (smashbot_state.action == Action.TURNING and smashbot_state.action_frame in range(2,12)): #
            if controller.prev.button[Button.BUTTON_Y]:
                controller.empty_input()
                return
            self.interruptible = False
            controller.press_button(Button.BUTTON_Y)
            return

        # Jump out of shine
        if isInShineStart:
            self.interruptible = False
            if smashbot_state.action_frame >= 3:
                controller.press_button(Button.BUTTON_Y)
                return
            else:
                controller.empty_input()
                return

        # We shouldn't need these. It's just there in case we miss the knee bend somehow
        jumping = [Action.JUMPING_ARIAL_FORWARD, Action.JUMPING_ARIAL_BACKWARD]

        # Airdodge back down into the stage
        if jcshine or smashbot_state.action in jumping:
            self.interruptible = False
            controller.press_button(Button.BUTTON_L)
            # Always wavedash the direction opponent is moving
            opponentspeed = opponent_state.speed_x_attack + opponent_state.speed_ground_x_self
            direction = opponentspeed > 0
            onleft = smashbot_state.x < opponent_state.x
            if abs(opponentspeed) < 0.01:
                direction = onleft

            # Unless we're RIGHT on top of the edge. In which case the only safe wavedash is back on the stage
            edge_x = melee.stages.EDGE_GROUND_POSITION[gamestate.stage]
            if opponent_state.x < 0:
                edge_x = -edge_x
            edgedistance = abs(edge_x - smashbot_state.x)
            if edgedistance < 0.5:
                direction = smashbot_state.x < 0

            # If we're facing the edge, it's not safe to wavedash off the edge. Adjust the distance down
            if smashbot_state.facing == smashbot_state.x > 0:
                distance = max(edgedistance / 15, 1)

            # Normalize distance from (0->1) to (-0.5 -> 0.5)
            delta = (self.distance / 2) # 0->0.5
            if not direction:
                delta = -delta
            controller.tilt_analog(Button.BUTTON_MAIN, 0.5 + delta, .35)
            return

        # If we're sliding and have shined, then we're all done here
        if smashbot_state.action == Action.LANDING_SPECIAL: #removed and self.hasshined
            self.interruptible = True
            controller.empty_input()
            return

        if smashbot_state.action in [Action.SWORD_DANCE_4_MID_AIR, Action.SWORD_DANCE_4_LOW_AIR]:
            self.interruptible = False
        else:
            self.interruptible = True

        controller.empty_input()
